spec/index.bs

```diff
@@ -1191,131 +1209,176 @@ access to it before garbage collection by calling {{GPUBuffer/destroy()}}.
 Note: This allows the user agent to reclaim the GPU memory associated with the {{GPUBuffer}}
 once all previously submitted operations using it are complete.
 
-### <dfn method for=GPUBuffer>destroy()</dfn> ### {#buffer-destroy}
+### <dfn method for=GPUBuffer>destroy()</dfn> ### {#GPUBuffer-destroy}
 
 <div algorithm="GPUBuffer.destroy">
-    <!-- TODO(kangz) handle error buffers once we have a description of the error monad -->
-    1. If the {{[[state]]}} slot of |this| is [=buffer state/mapped for reading=] or [=buffer state/mapped for writing=]:
+  **Arguments:** (None)
 
-        1. Run the steps to unmap `"this"`
+  **Returns:** (None)
 
-    1. Set the {{[[state]]}} slot of |this| to [=buffer state/destroyed=]
-</div>
+  1. If the {{[[state]]}} slot of |this| is [=buffer state/mapped=] or [=buffer state/mapped at creation=]:
 
-## Buffer Usage ## {#buffer-usage}
+    1. Run the steps to unmap `"this"`
 
-<script type=idl>
-typedef [EnforceRange] unsigned long GPUBufferUsageFlags;
-interface GPUBufferUsage {
-    const GPUBufferUsageFlags MAP_READ      = 0x0001;
-    const GPUBufferUsageFlags MAP_WRITE     = 0x0002;
-    const GPUBufferUsageFlags COPY_SRC      = 0x0004;
-    const GPUBufferUsageFlags COPY_DST      = 0x0008;
-    const GPUBufferUsageFlags INDEX         = 0x0010;
-    const GPUBufferUsageFlags VERTEX        = 0x0020;
-    const GPUBufferUsageFlags UNIFORM       = 0x0040;
-    const GPUBufferUsageFlags STORAGE       = 0x0080;
-    const GPUBufferUsageFlags INDIRECT      = 0x0100;
-    const GPUBufferUsageFlags QUERY_RESOLVE = 0x0200;
-};
-</script>
+  1. Set the {{[[state]]}} slot of |this| to [=buffer state/destroyed=]
+
+  Issue: Handle error buffers once we have a description of the error monad.
+</div>
 
 ## Buffer Mapping ## {#buffer-mapping}
 
-An application can request to map a {{GPUBuffer}} to get its mapping which is
-an {{ArrayBuffer}} representing the {{GPUBuffer}}'s allocation. Mappings are
-requested asynchronously so that the user agent can ensure the GPU finished
-using the {{GPUBuffer}} before the application gets its mapping. Mappings can
-be requested for reading with {{GPUBuffer/mapReadAsync}} or writing with
-{{GPUBuffer/mapWriteAsync}}. A mapped {{GPUBuffer}} cannot be used by the GPU
-and must be unmapped using {{GPUBuffer/unmap}} before it can be used on the
-[=Queue timeline=].
+An application can request to map a {{GPUBuffer}} so that they can access its
+content via {{ArrayBuffer}}s that represent part of the {{GPUBuffer}}'s
```

Comment:
What happens if I create a buffer with MAP_READ, I map it, write some data to it, then unmap, then map again?
The spec says, if I read this correctly, that the next mapping should get me the GPU contents of the buffer, not whatever I wrote to it, right? How would this be achieved?

Reply:
Your reading of the spec if correct. `MAP_READ` would always be staged and perform a copy of the actual buffer content to memory visible to JS through an `ArrayBuffer`. This is optimal for implementations that can't wrap shmem in a GPU resource but suboptimal for those that can.

There's a way to achieve zero-copy while keeping portable semantics (whether we can wrap shmem or not). Implementations wrapping shmem would have no special logic. Implementations not wrapping shmem would have to keep track of the exact ranges of `MAP_READ` buffers that are touched by the GPU (taking into account untouched bytes because of row pitch for example), and on `mapAsync` send the list of dirty regions with the updated content. This interacts badly with the ability to map only a subrange because you could only flush the dirty regions for the mapped subrange (or send more info than what's for the subrange).

Alternatively non-wrapping implementations could send back the new data on `unmap` so it could be updated in the GPU process.

The issue is that contrary to languages like C, C++ or Rust, you can't give readonly access to data in Javascript so either we make `MAP_READ` truly readonly by doing some things suboptimally, or we make `MAP_READ` a `MAP_READWRITE_BUT_ONLY_ALSO_COPY_DST` that incurs different complexity and overhead.

Reply:
My concern is that the implementation wouldn't even be able to track ranges changed by GPU. For example, what if the user changed byte 5 (in the ArrayBuffer that is supposedly read-only), GPU didn't touch it, and the user requests to map it again? The implementation would still have to invalidate that chunk by re-reading from GPU, even though GPU didn't touch it.

I guess it's still doable with a bit of complexity: consider any ranges that are given to the user to be dirty for the purposes of invalidation. Anyway, this needs to be specified (the behavior in case the user writes to the ArrayBuffer).

Reply:
I think it is already specified because:

 - When the promise for mapAsync resolves, it creates and `ArrayBuffer` `m` and sets its content to the GPU's content.
 - On unmap, **only if mapped at creation or `MAP_WRITE`** then the content of the buffer is updated.

I added a note that calls this out.

Reply:
Yes, the spec has this right. My question is about how you are suggesting the implementations to behave in case the ArrayBuffer wraps a shared memory block, and the user decides to write into it even though it's MAP_READ.

Reply:
I suggest that for `MAP_READ` implementation don't wrap shmem in a GPU resource, and always perform a copy from the GPU process readback buffer to the shmem. This way modifications written in the `ArrayBuffer` are always overwritten.